product_variant_default_code/models/product.py

```diff
@@ -0,0 +1,206 @@
+# -*- coding: utf-8 -*-
+# Copyright 2014 AvancOSC - Alfredo de la Fuente
+# Copyright 2014 Tecnativa - Pedro M. Baeza
+# Copyright 2014 Shine IT - Tony Gu
+# Copyright 2017 Tecnativa - David Vidal
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+import re
+from string import Template
+from collections import defaultdict
+
+
+class ReferenceMask(Template):
+    pattern = r'''\[(?:
+                    (?P<escaped>\[) |
+                    (?P<named>[^\]]+?)\] |
+                    (?P<braced>[^\]]+?)\] |
+                    (?P<invalid>)
+                    )'''
+
+
+def extract_token(s):
+    pattern = re.compile(r'\[([^\]]+?)\]')
+    return set(pattern.findall(s))
+
+
+def sanitize_reference_mask(product, mask):
+    tokens = extract_token(mask)
+    attribute_names = set()
+    for line in product.attribute_line_ids:
+        attribute_names.add(line.attribute_id.name)
+    if not tokens.issubset(attribute_names):
+        raise UserError(_('Found unrecognized attribute name in "Variant '
+                          'Reference Mask"'))
+
+
+def get_rendered_default_code(product, mask):
+    product_attrs = defaultdict(str)
+    reference_mask = ReferenceMask(mask)
+    for value in product.attribute_value_ids:
+        if value.attribute_id.code:
+            product_attrs[value.attribute_id.name] += value.attribute_id.code
+        if value.code:
+            product_attrs[value.attribute_id.name] += value.code
+    all_attrs = extract_token(mask)
+    missing_attrs = all_attrs - set(product_attrs.keys())
+    missing = dict.fromkeys(
+        missing_attrs,
+        product.env['ir.config_parameter'].get_param(
+            'default_reference_missing_placeholder'))
+    product_attrs.update(missing)
+    default_code = reference_mask.safe_substitute(product_attrs)
+    return default_code
+
+
+def render_default_code(product, mask):
+    product.default_code = get_rendered_default_code(product, mask)
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    prefix_code = fields.Char(
+        string='Prefix code',
+        help='Add prefix to product variant default code',
+    )
+    reference_mask = fields.Char(
+        string='Variant reference mask',
+        help='Reference mask for building internal references of a '
+             'variant generated from this template.\n'
+
```

Comment:
Is this field support translations ? ie:
user A is english and has `Color` attribute
user B is french and has the same attribute but named `Couleur`

It seems it break the behavior:
- create a product in english with Color attribute
- switch user to french : it display correctly with 'Color' attribute in mask
- try to duplicate: "Found unrecognized attribute name in "Variant Reference Mask" because it search Couleur instead of Color
- if you create a new product now 'Color' doesn't work :-(

I think we shouldn't use a translatable field as ref in mask field: either attribute code, either a new field untranslatable (ie 'color' english version)

Reply:
OK for not being translatable.